### PR TITLE
Restructure the `org-permissions` guard to be Angular 17+ compliant

### PR DESCRIPTION
**Pull request by addisonbeck**
# Restructure the `org-permissions` guard to be Angular 17+ compliant

## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/AC-2759

## 📔 Objective

In the web vault: in order to support Angular versions 17 and higher we must
restructure the way organization landing page URL redirects are written and
configured.

Currently most URL guards in the Bitwarden Angular clients are written as
router guard classes. The base `CanActivate` / `CanDeactivate` classes have
been deprecated by the Angular team and are subject to removal. See [the](https://angular.io/guide/deprecations#router-class-and-injectiontoken-guards-and-resolvers)
[Angular deprecation notice](https://angular.io/guide/deprecations#router-class-and-injectiontoken-guards-and-resolvers) for details.

At the time of writing, platform team has already completed this work for
their guards [here](https://bitwarden.atlassian.net/browse/PM-4349). Any guard declarations are swapped from exporting a
class to exporting a function, and the function is added to CanActivate
arrays in routers the same way the classes already are. There are some other
guards that are already function based scattered throughout the codebase that
can also be used as examples for style, like [auth team's lock guard](https://github.com/bitwarden/clients/blob/1aaa88a64d4d48aa75868980eb4c280b672057c3/libs/angular/src/auth/guards/lock.guard.ts).

This PR achieves this goal for the `org-permissions` guard.

This change should result in no functional changes to the system. It is
purely a code restructure to comply with a deprecation in Angular.

This ticket applies to urls that are:
1. Subroutes of `<https://THE_WEB_VAULT/organizations/{id}/>`
1. `<https://THE_WEB_VAULT/sm/settings/import>`
1. `<https://THE_WEB_VAULT/sm/settings/export>`

> ℹ️ There are many subroutes of organizations/{id}, too many to list at
> once and test individually. If specific examples are needed for testing
> please reach out to development.

For URLs that are subject to this guard the following rules apply:
1. If the organizationId route parameter is not set or is not valid for the
user then the user is redirected to /
1. There is a special case to support linkable ciphers. If the user fails the
permission check, meaning they can not access the admin console, but a cipher
id is found in the route parameters the user is redirected to /vault
with the item id passed along.
1. If the required permissions are found navigation proceeds as normal. If the
required permissions are not found the user should be redirected to
/organizations/{id} or / depending on whether or not they have access to
the admin console at all.

### Organization ID Not Found

In this example I try to access an organization ID that isn't found in the logged in user's organizations, and am redirected back to the home screen

https://github.com/bitwarden/clients/assets/15897251/29d70801-30fa-4f2a-bd1d-e135622261f5

### Linkable Ciphers

In this example I have view access to a cipher that I have a link to from the admin console. Instead of breaking, or being send to the admin console, I am redirected to the cipher in my vault.

https://github.com/bitwarden/clients/assets/15897251/f98d851d-5982-41a1-b9ae-adf602f248f7

### Permission Denied

In this example I am an organization user that can manage groups, but I try to access the members screen from the URL bar and am blocked.

https://github.com/bitwarden/clients/assets/15897251/81d4117f-d793-4a64-9824-0171895718c5

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts [hunk 1]
```diff
@@ -10,40 +15,51 @@ import { I18nService } from "@bitwarden/common/platform/abstractions/i18n.servic
...
     }
 
     if (!org.isOwner && !org.enabled) {
-      this.platformUtilsService.showToast(
-        "error",
-        null,
-        this.i18nService.t("organizationIsDisabled"),
-      );
-      return this.router.createUrlTree(["/"]);
+      platformUtilsService.showToast("error", null, i18nService.t("organizationIsDisabled"));
```
Comment (eliykat) [inline apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts:59]: While we're here, please replace this with the newer `ToastService` from the CL.

File: apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts [hunk 2]
```diff
@@ -52,19 +68,19 @@ export class OrganizationPermissionsGuard implements CanActivate {
...
-        return this.router.createUrlTree(["/vault"], {
+        return router.createUrlTree(["/vault"], {
           queryParams: {
             itemId: cipherId,
           },
         });
       }
 
-      this.platformUtilsService.showToast("error", null, this.i18nService.t("accessDenied"));
+      platformUtilsService.showToast("error", null, i18nService.t("accessDenied"));
```
Comment (eliykat) [inline apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts:78]: And here

File: apps/web/src/app/admin-console/organizations/reporting/organization-reporting-routing.module.ts [hunk 3]
```diff
@@ -84,10 +83,9 @@ const routes: Routes = [
       {
         path: "events",
         component: EventsComponent,
-        canActivate: [OrganizationPermissionsGuard],
+        canActivate: [organizationPermissionsGuard((org: Organization) => org.canAccessEventLogs)],
```
Comment (eliykat) [inline apps/web/src/app/admin-console/organizations/reporting/organization-reporting-routing.module.ts:86]: nit: explicit typing of `Organization` is probably unnecessary here now that the function parameter is typed. (And in other locations - doesn't really matter though, your call.)

File: apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.spec.ts [hunk 4]
```diff
@@ -47,19 +46,23 @@ describe("Organization Permissions Guard", () => {
       },
     });
 
-    organizationPermissionsGuard = new OrganizationPermissionsGuard(
```
Comment (eliykat) [inline apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.spec.ts:50]: Above this, line ~45: old reference to `data.organizationPermissions`.

**Review by eliykat — CHANGES_REQUESTED** [review]
Great job with these @addisonbeck 👏

File: apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts (hunk 1 above)
Comment (addisonbeck) [reply apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts]: Implemented as suggested on https://github.com/bitwarden/clients/pull/9631/commits/30a200b6c8d581b15a4f7df524c51ccf28e85c8c

File: apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts (hunk 2 above)
Comment (addisonbeck) [reply apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.ts]: Implemented as suggested on https://github.com/bitwarden/clients/pull/9631/commits/30a200b6c8d581b15a4f7df524c51ccf28e85c8c

File: apps/web/src/app/admin-console/organizations/reporting/organization-reporting-routing.module.ts (hunk 3 above)
Comment (addisonbeck) [reply apps/web/src/app/admin-console/organizations/reporting/organization-reporting-routing.module.ts]: No complaints from me, I agree this level of verbosity is not really needed. Implemented as suggested on https://github.com/bitwarden/clients/pull/9631/commits/8d98a93bf758e1d8415ef10d3febfb35c80bdd4f

File: apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.spec.ts (hunk 4 above)
Comment (addisonbeck) [reply apps/web/src/app/admin-console/organizations/guards/org-permissions.guard.spec.ts:50]: Removed on https://github.com/bitwarden/clients/pull/9631/commits/caa1b7aad302a7f8667d2c68959393b916df7b22

**Review by eliykat — APPROVED** [review]

**Review by Thomas-Avery — APPROVED** [review]
SM changes LGTM

**Review by amorask-bitwarden — DISMISSED** [review]

**Review by shane-melton — APPROVED** [review]
Vault change looks good!

**Comment by djsmith85** [comment]
@addisonbeck Conflict was created by merging https://github.com/bitwarden/clients/pull/9741:

`apps/web/src/app/admin-console/organizations/tools/vault-export/org-vault-export-routing.module.ts` has been deleted and the "export"-route has been moved into `apps/web/src/app/admin-console/organizations/settings/organization-settings-routing.module.ts`

**Review by eliykat — APPROVED** [review]

**Comment by addisonbeck** [comment]
@eliykat @shane-melton @amorask-bitwarden @Thomas-Avery Hey folks! This passed QA and I'm ready to get it merged, if you wouldn't mind giving it another stamp. Nothing changed during QA besides keeping the branch up to date with `main`. Thank you!

**Review by cturnbull-bitwarden — APPROVED** [review]

**Pull request merged**